merlion/models/forecast/deep_models/deep_forecast_base.py

```diff
@@ -0,0 +1,340 @@
+        else:
+            raise NotImplementedError("%s optimizer is not supported yet..." % (self.config.optim_name))
+
+        # initialize loss function
+        if self.config.criterion == "mse":
+            self.loss_fn = nn.MSELoss()
+        else:
+            raise NotImplementedError("%s loss is not supported..." % (self.config.criterion))
+
+        self._post_model_creation()
+
+    def _post_model_creation(self):
+        pass
+
+    def evaluate(self, evaluate_data: pd.DataFrame, metric: str = "mse"):
+        config = self.config
+        self.deep_model.eval()
+        eval_dataset = RollingWindowDataset(
+            evaluate_data,
+            n_past=config.n_past,
+            n_future=config.max_forecast_steps,
+            batch_size=config.batch_size,
+            target_seq_index=None,  # have to set None, we use config.target_seq_index later in the training
+            ts_encoding=config.ts_encoding,
+            ts_freq=config.ts_freq,
+            start_token_len=config.start_token_len,
+            flatten=False,
+            shuffle=False,
+        )
+
+        all_preds = []
+        all_trues = []
+
+        for i, batch in enumerate(eval_dataset):
+            with torch.no_grad():
+                loss, outputs, y_true = self._deep_batch_iter(batch, self.config)
+                pred = outputs.detach().cpu().numpy()
+                true = y_true.detach().cpu().numpy()
+                all_preds.append(pred)
+                all_trues.append(true)
+
+        preds = np.concatenate(all_preds, axis=0)
+        trues = np.concatenate(all_trues, axis=0)
+
+        logger.info("test shape:" + str(preds.shape) + str(trues.shape))
+
+        if metric == "mse":
+            pred_err = np.mean(np.sum((preds - trues) ** 2, axis=-1), axis=-1)
+        else:
+            raise NotImplementedError
+
+        return preds, pred_err
+
+    def train_pre_process(
+        self, train_data: TimeSeries, exog_data: TimeSeries = None, return_exog=None
+    ) -> Union[TimeSeries, Tuple[TimeSeries, Union[TimeSeries, None]]]:
+        train_data = super(ForecasterBase, self).train_pre_process(train_data)
+
+        if self.dim == 1:
+            self.config.target_seq_index = 0
+        if self.config.target_seq_index is None:
+            self.target_name = "Multi dimensional forecasting"
+        else:
+            assert 0 <= self.config.target_seq_index < train_data.dim, (
+                f"Expected `target_seq_index` to be between 0 and {train_data.dim} "
+                f"(the dimension of the transformed data), but got {self.config.target_seq_index}"
+            )
+
+            self.target_name = train_data.names[self.target_seq_index]
+
+        return (train_data, exog_data) if return_exog else train_data
+
+    def _train(self, train_data: pd.DataFrame, train_config=None) -> pd.DataFrame:
+
+        if train_config is not None:
+            self.config = copy.deepcopy(train_config)
+        config = self.config
```

Comment:
I'd strongly advise against doing this. This can have a lot of unintended side effects, since parts of the config are initialized by `train_pre_process()`. If you really want, you can consider decomposing `lr`, `loss_fn`, and some other parameters which only affect the training procedure from the `DeepConfig` into a `train_config`. But in my opinion, this is probably unnecessary.